File: grind75/Jan02/139.go
func wordBreak(s string, wordDict []string) bool {
    m := map[string]bool{}
    var dfs func(string) bool
    dfs = func (cur string) bool {
        k := cur
        if cur == s {
            return true
        }
        if r, ok := m[k]; ok {
            return r
        }
        for i := 0; i < len(wordDict); i++ {
            if !strings.HasPrefix(s, cur+wordDict[i]) {
                continue
            }
            if dfs(cur+wordDict[i]) {
                m[k] = true
                return true
            }
        }
        m[k] = false
        return false
    }
    return dfs("")
}

File: grind75/Jan02/139.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        # knapsack problem
        dp = [False] * (len(s) + 1)
        dp[0] = True
        for i in range(len(s)+1):
            for word in wordDict:
                n = len(word)
                if i - n >= 0 and dp[i-n] and s[i-n:i] == word:
                    dp[i] = True
        
        return dp[-1]
